### PR TITLE
Add search query and citations to the chat python SDK 

### PR DESCRIPTION
**Pull request by BeatrixCohere**
Close: PTS-4743

This adds:
- Search query only parameter
- Chat with documents 
- Chat with connectors
- Citation quality parameter

### CODE REVIEW

File: cohere/client.py [hunk 1]
```diff
@@ -265,6 +269,25 @@ def chat(
...
+                        "url": "https://education.nationalgeographic.org/resource/mount-everest/",
+                    },
+                    {
+                        "id": "national_geographic_mariana",
+                        "title": "Depth of the Mariana Trench",
+                        "snippet": "The depth of the Mariana Trench is 36,070 feet",
+                        "url": "https://www.nationalgeographic.org/activity/mariana-trench-deepest-place-earth",
+                    },
+                ],
+            connectors (List[Dict[str, str]]): (Optional) When specified, the model's reply will be enriched with information found by quering each of the connectors (RAG). Example: connectors=[{"id": "web-search"}]
```
Comment (sanderland) [inline cohere/client.py:289]: For connectors, how would someone find the allowed values?

File: cohere/client.py [hunk 2]
```diff
@@ -314,7 +370,12 @@ def chat(
             "p": p,
             "k": k,
             "logit_bias": logit_bias,
+            "search_queries_only": search_queries_only,
+            "documents": documents,
+            "connectors": connectors,
         }
+        if citation_quality is not None:
+            json_body["citation_quality"] = citation_quality
```
Comment (sanderland) [inline cohere/client.py:378]: does the API not like None as default here?

File: cohere/responses/chat.py [hunk 3]
```diff
@@ -152,6 +216,11 @@ def _make_response_item(self, index, line) -> Any:
             self.chatlog = response.get("chatlog")
             self.token_count = response.get("token_count")
             self.meta = response.get("meta")
+            self.is_search_required = (response.get("is_search_required"),)  # optional
```
Comment (sanderland) [inline cohere/responses/chat.py:219]: why the tuples here?

**Review by sanderland — COMMENTED** [review]
Looks great. Mainly concerned about the single item tuples being wrong.

**Review by harry-cohere — APPROVED** [review]

File: cohere/client.py (hunk 1 above)
Comment (BeatrixCohere) [reply cohere/client.py:289]: For the short term, this will be documented. For the longer term, there will be a call to the API to list connectors.

**Review by sanderland — APPROVED** [review]

File: cohere/responses/chat.py [hunk 4]
```diff
@@ -75,16 +91,27 @@ async def respond(self, response: str, max_tokens: int = None) -> "AsyncChat":
         )
 
 
+class Events(str, Enum):
```
Comment (mkozakov) [inline cohere/responses/chat.py:94]: Should be singular `StreamEvent`?

File: cohere/responses/chat.py [hunk 5]
```diff
@@ -75,16 +91,27 @@ async def respond(self, response: str, max_tokens: int = None) -> "AsyncChat":
...
+    TEXT_GENERATION = "text-generation"
+    CITATION_GENERATION = "citation-generation"
+    STREAM_END = "stream-end"
+
+
 class StreamResponse(CohereObject):
     def __init__(
         self,
         is_finished: bool,
+        event_type: str,
```
Comment (mkozakov) [inline cohere/responses/chat.py:107]:
```suggestion
        event_type: typing.Union([StreamEvent, str])
```

File: cohere/client.py [hunk 6]
```diff
@@ -296,6 +319,39 @@ def chat(
                 >>>     return_prompt=True)
                 >>> print(res.text)
                 >>> print(res.prompt)
+            Generate search queries for fetching documents to use in chat:
+                >>> res = co.chat(
+                >>>     "What is the height of Mount Everest?",
+                >>>      search_queries_only=True)
```
Comment (mkozakov) [inline cohere/client.py:325]: up to you, but this is the most niche use case, so imo probably should be at the bottom.

**Pull request merged**